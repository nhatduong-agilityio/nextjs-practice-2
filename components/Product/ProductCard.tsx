// Libs
import React, { Fragment, memo, useState } from 'react'
import {
  Box,
  Card as ChakraCard,
  CardBody,
  CardHeader,
  Flex,
  Heading,
  IconButton,
  Text,
} from '@chakra-ui/react'
import NextLink from 'next/link'

// Components
import HeartIcon from '@components/Icons/HeartIcon'
import StarIcon from '@components/Icons/StarIcon'
import Image from '@components/Common/Image'

// Types
import type { TProduct } from '@matched-types/product'
import { PAGE_URL } from '@constants/routes'

interface IProps {
  product: TProduct
  variant?: 'small' | 'medium' | 'large'
  headingCard?: 'small' | 'medium' | 'large'
  sizeImage?: 'small' | 'medium'
}

const RATING = 5

const ProductCard = memo(
  ({ product, variant = 'large', headingCard = 'large', sizeImage = 'medium' }: IProps) => {
    const [wishList, setWishList] = useState(false)

    // valid sales percent
    const validDiscount = product.discount > 0 && product.discount <= 100

    // calculator price
    const salePrice = product.price - (product.price * product.discount) / 100

    const handleWishList = () => {
      setWishList(!wishList)
    }

    return (
      <ChakraCard boxShadow='none'>
        <CardHeader p={0} flexDir='row'>
          <Box
            position='absolute'
            zIndex='2'
            w='full'
            display='flex'
            justifyContent='end'
            p={['10px 15px', '15px 15px']}
          >
            <IconButton
              aria-label='wishlist-button'
              borderRadius='50px'
              minW='30px'
              width={['30px', '40px']}
              height={['30px', '40px']}
              onClick={handleWishList}
            >
              <HeartIcon
                width={['30px', '40px']}
                height={['30px', '40px']}
                viewBox='0 0 30 30'
                fill={wishList ? 'red' : '#C0C0CD'}
                stroke={wishList ? 'red' : '#C0C0CD'}
              />
            </IconButton>
          </Box>
          <NextLink href={`${PAGE_URL.PRODUCT.URL}/${product.id}`} passHref>
            <Image
              width='full'
              alt={product.imageAlt}
              src={product.images.image}
              height={['150', `${sizeImage === 'medium' ? '300' : '259'}`]}
              variant='primary'
            />
          </NextLink>
        </CardHeader>
        <CardBody p={['10px 0 5px 0', '15px 0 10px 0', '15px 0 10px 0']}>
          <Box>
            <NextLink href={`${PAGE_URL.PRODUCT.URL}/${product.id}`} passHref>
              <Heading variant={['small', `${headingCard}`]} lineHeight='normal'>
                {product.name}
              </Heading>
            </NextLink>
            <Box display='flex' pt={['5px', '10px', '10px']} alignItems='center'>
              {Array(RATING)
                .fill('')
                .map((_, index) => (
                  // eslint-disable-next-line react/no-array-index-key
                  <Fragment key={index}>
                    <StarIcon
                      width={['15px', '18px', '18px']}
                      height={['15px', '18px', '18px']}
                      viewBox='0 0 15 15'
                      fill={index < product.rating ? '#F9BB1D' : 'gray.300'}
                      style={{ marginRight: '1px' }}
                    />
                  </Fragment>
                ))}
              <Text ml='2' size={['14', `${variant === 'medium' ? 14 : 16}`]} variant='grayScheme'>
                ({product.sellCount} Sell)
              </Text>
            </Box>
            <Flex flexDir='row' alignItems='center' w='full' gap={['5px', '10px']}>
              <Text
                mr='10px'
                noOfLines={1}
                size={[`${variant === 'medium' ? 16 : 18}`, `${variant === 'medium' ? 22 : 25}`]}
                variant='large'
              >
                ${salePrice}
              </Text>
              {validDiscount && (
                <>
                  <Text
                    as='del'
                    mr='3px'
                    noOfLines={1}
                    size={[
                      `${variant === 'medium' ? 16 : 18}`,
                      `${variant === 'medium' ? 15 : 25}`,
                    ]}
                    variant='grayScheme'
                  >
                    ${product.price}
                  </Text>
                  <Text
                    variant='redScheme'
                    noOfLines={1}
                    size={[
                      `${variant === 'medium' ? 16 : 18}`,
                      `${variant === 'medium' ? 15 : 25}`,
                    ]}
                  >
                    -{product.discount}%
                  </Text>
                </>
              )}
            </Flex>
          </Box>
        </CardBody>
      </ChakraCard>
    )
  },
)

ProductCard.displayName = 'ProductCard'

export default ProductCard
