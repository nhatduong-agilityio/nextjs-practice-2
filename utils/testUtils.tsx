import { ChakraProvider } from '@chakra-ui/react'
import { AuthProvider } from '@store/providers/AuthProvider'
import { LoadingProvider } from '@store/providers/LoadingProvider'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'
import { customTheme } from '@themes/index'
import type { ReactElement } from 'react'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={customTheme}>
      <LoadingProvider>
        <AuthProvider>{children}</AuthProvider>
      </LoadingProvider>
    </ChakraProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>): RenderResult =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
