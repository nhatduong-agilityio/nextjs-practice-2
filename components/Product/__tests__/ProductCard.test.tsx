// Utils
import { render } from '@utils/testUtils'

import { ProductsMock } from '@mock/dataMock'

// Components
import type { TProduct } from '@matched-types/product'
import ProductCard from '../ProductCard'

describe('ProductCard render', () => {
  const propsMock = {
    product: ProductsMock[0] ? ProductsMock[0] : ({} as TProduct),
  }

  it('Should ProductCard match Snapshot', () => {
    const { container } = render(<ProductCard {...propsMock} />)

    expect(container).toMatchSnapshot()
  })
})
