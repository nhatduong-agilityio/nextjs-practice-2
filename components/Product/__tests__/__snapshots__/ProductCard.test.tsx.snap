// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard render Should ProductCard match Snapshot 1`] = `
<div>
  <div
    class="chakra-card css-1fnaag1"
  >
    <div
      class="chakra-card__header css-e3ulnx"
    >
      <div
        class="css-rszsdl"
      >
        <button
          aria-label="wishlist-button"
          class="chakra-button css-jvwple"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-cjaec7"
            focusable="false"
            viewBox="0 0 30 30"
          >
            <circle
              cx="15"
              cy="15"
              fill="white"
              r="15"
              stroke-width="0"
            />
            <g
              clip-path="url(#clip0_187_701)"
            >
              <path
                d="M12.5 9.0625L10.625 9.6875L8.75 10.9375V15L12.1875 18.4375L15 20.625L19.375 16.875L20.9375 14.375V11.5625L19.6875 9.6875L17.1875 9.375L15.3125 10.9375L14.6875 10.3125L12.5 9.0625Z"
              />
              <path
                d="M15.7805 9.39434C15.4958 9.61931 15.2349 9.8863 15 10.1914C14.7652 9.88628 14.5042 9.61931 14.2196 9.39436L14.2196 9.39435C13.5541 8.86841 12.7946 8.6 11.9662 8.6C10.8617 8.6 9.84462 9.05247 9.10443 9.8728L9.10442 9.87281C8.37363 10.6828 7.975 11.7848 7.975 12.9724C7.975 14.2028 8.4256 15.3228 9.37225 16.49L9.37226 16.49C10.2129 17.5265 11.4176 18.5755 12.7962 19.776L12.8029 19.7818L12.804 19.7828L12.8053 19.7839L12.8053 19.7839L12.8065 19.7849C13.279 20.1964 13.8138 20.6621 14.3684 21.1577C14.5427 21.3137 14.767 21.4 15 21.4C15.2329 21.4 15.4572 21.3137 15.6314 21.1579C16.1873 20.6612 16.723 20.1947 17.1964 19.7826L17.1964 19.7826L17.2053 19.7748C18.5833 18.5748 19.7874 17.5262 20.6278 16.4899C21.5745 15.3228 22.025 14.2028 22.025 12.9724C22.025 11.7849 21.6265 10.6828 20.8957 9.87281L20.8957 9.8728C20.1555 9.05247 19.1383 8.6 18.0339 8.6C17.2055 8.6 16.446 8.86841 15.7805 9.39434ZM15.7805 9.39434L15.8735 9.51204M15.7805 9.39434L15.8735 9.51204M15.8735 9.51204C16.5133 9.00641 17.2402 8.75 18.0339 8.75C19.0958 8.75 20.0725 9.18445 20.7843 9.97329L15.0933 10.3165C15.3278 10.0063 15.5889 9.73699 15.8735 9.51204ZM9.91923 10.6316C10.4483 10.0454 11.1742 9.72303 11.9662 9.72303C12.5423 9.72303 13.0709 9.90941 13.5408 10.2807C13.9623 10.6137 14.2586 11.0376 14.4337 11.3376L14.4338 11.3376C14.5528 11.5414 14.7648 11.6655 15 11.6655C15.2352 11.6655 15.4472 11.5414 15.5662 11.3376L15.4367 11.262L15.5662 11.3376C15.7415 11.0376 16.0378 10.6138 16.4592 10.2807C16.9291 9.90942 17.4577 9.72303 18.0339 9.72303C18.8258 9.72303 19.5518 10.0454 20.0808 10.6316C20.6187 11.2278 20.9196 12.0576 20.9196 12.9724C20.9196 13.9284 20.5745 14.7898 19.7748 15.7759C18.9971 16.7348 17.837 17.7455 16.477 18.93C16.0299 19.3191 15.5264 19.7576 14.999 20.2258C14.4746 19.7585 13.9719 19.3207 13.5259 18.9323L13.5246 18.9313L13.5234 18.9302C13.5232 18.93 13.523 18.9298 13.5227 18.9297C12.1628 17.7454 11.0029 16.7347 10.2253 15.7759L10.2253 15.7759C9.42554 14.7898 9.08035 13.9284 9.08035 12.9724C9.08035 12.0576 9.3813 11.2278 9.91923 10.6316ZM9.91923 10.6316L9.80787 10.5311M9.91923 10.6316L9.80787 10.5311M9.80787 10.5311C10.3654 9.91326 11.1319 9.57303 11.9662 9.57303C12.5772 9.57303 13.1384 9.77155 13.6338 10.163M9.80787 10.5311C9.24202 11.1583 8.93035 12.0253 8.93035 12.9724C8.93035 13.9718 9.29384 14.8656 10.1088 15.8704L13.6338 10.163M13.6338 10.163C14.0755 10.512 14.3831 10.9532 14.5633 11.262L13.6338 10.163ZM11.9662 8.75C12.7599 8.75 13.4868 9.00641 14.1266 9.51204L12.9025 19.6697C11.5173 18.4634 10.3211 17.4217 9.48875 16.3955C8.55836 15.2484 8.125 14.1607 8.125 12.9724C8.125 11.818 8.51241 10.7529 9.21579 9.97329C9.92757 9.18445 10.9042 8.75 11.9662 8.75Z"
                stroke-width="0.3"
              />
            </g>
            <defs>
              <clippath
                id="clip0_187_701"
              >
                <rect
                  fill="white"
                  height="15"
                  transform="translate(7.5 7.5)"
                  width="15"
                />
              </clippath>
            </defs>
          </svg>
        </button>
      </div>
      <a
        href="/product/1146ea86-4853-4e7a-8d9a-d6b720902d79"
      >
        <div
          class="css-q53xaa"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="Picture of Full set Grey Sofa With Pillows"
              data-nimg="fill"
              decoding="async"
              sizes="100vw"
              src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
            />
          </span>
        </div>
      </a>
    </div>
    <div
      class="chakra-card__body css-f99mjs"
    >
      <div
        class="css-0"
      >
        <a
          href="/product/1146ea86-4853-4e7a-8d9a-d6b720902d79"
        >
          <h2
            class="chakra-heading css-10zd33"
          >
            Green Slepper Sofa
          </h2>
        </a>
        <div
          class="css-1hzkocu"
        >
          <svg
            class="chakra-icon css-tlkdll"
            focusable="false"
            style="margin-right: 1px;"
            viewBox="0 0 15 15"
          >
            <path
              d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
            />
          </svg>
          <svg
            class="chakra-icon css-tlkdll"
            focusable="false"
            style="margin-right: 1px;"
            viewBox="0 0 15 15"
          >
            <path
              d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
            />
          </svg>
          <svg
            class="chakra-icon css-tlkdll"
            focusable="false"
            style="margin-right: 1px;"
            viewBox="0 0 15 15"
          >
            <path
              d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
            />
          </svg>
          <svg
            class="chakra-icon css-tlkdll"
            focusable="false"
            style="margin-right: 1px;"
            viewBox="0 0 15 15"
          >
            <path
              d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
            />
          </svg>
          <svg
            class="chakra-icon css-tlkdll"
            focusable="false"
            style="margin-right: 1px;"
            viewBox="0 0 15 15"
          >
            <path
              d="M7.5 10.7938L11.3625 13.125L10.3375 8.73125L13.75 5.775L9.25625 5.39375L7.5 1.25L5.74375 5.39375L1.25 5.775L4.6625 8.73125L3.6375 13.125L7.5 10.7938Z"
            />
          </svg>
          <p
            class="chakra-text css-1ma697p"
          >
            (
            400
             Sell)
          </p>
        </div>
        <div
          class="css-hdz8n1"
        >
          <p
            class="chakra-text css-13ubzgw"
          >
            $
            120
          </p>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
